File: LSQFit.py
import numpy as np
from numpy.linalg import inv
from matplotlib import pyplot as plt
from matplotlib.backends.backend_pdf import PdfPages
from math import log
from random import gauss

xmin=1.0
xmax=20.0
npoints=12       
sigma=0.2        
pars=[0.5,1.3,0.5]

lx=np.zeros(npoints)
ly=np.zeros(npoints)
ley=np.zeros(npoints)

def f(x,par):
    return par[0]+par[1]*log(x)+par[2]*log(x)*log(x)

def getX(x):
    step=(xmax-xmin)/npoints
    for i in range(npoints):
        x[i]=xmin+i*step
        
def getY(x,y,ey):
    for i in range(npoints):
        y[i]=f(x[i],pars)+gauss(0,sigma)
        ey[i]=sigma

# --- least squares fit ---
def fit_once():
    getX(lx)
    getY(lx,ly,ley)
    A = np.zeros((npoints,3))
    for i in range(npoints):
        L = log(lx[i])
        A[i,0] = 1
        A[i,1] = L
        A[i,2] = L*L
    
    W = np.diag(1.0/ley**2)
    AtW = A.T @ W
    cov = inv(AtW @ A)
    p = cov @ (AtW @ ly)
    
    chi2 = np.sum(((ly - A @ p)/ley)**2)
    ndf = npoints - 3
    return p, chi2, chi2/ndf

# --- run experiments ---
nexperiments = 2000

par_a = []
par_b = []
par_c = []
chi2_vals = []
chi2_red  = []

for _ in range(nexperiments):
    p, c2, c2r = fit_once()
    par_a.append(p[0])
    par_b.append(p[1])
    par_c.append(p[2])
    chi2_vals.append(c2)
    chi2_red.append(c2r)

par_a = np.array(par_a)
par_b = np.array(par_b)
par_c = np.array(par_c)
chi2_vals = np.array(chi2_vals)
chi2_red  = np.array(chi2_red)

# --- first 2x2: a, b, c, chi2 ---
fig1, ax1 = plt.subplots(2,2, figsize=(8,8))
ax1[0,0].hist(par_a, bins=40)
ax1[0,0].set_title("a distribution")
ax1[0,1].hist(par_b, bins=40)
ax1[0,1].set_title("b distribution")
ax1[1,0].hist(par_c, bins=40)
ax1[1,0].set_title("c distribution")
ax1[1,1].hist(chi2_vals, bins=40)
ax1[1,1].set_title("Chi2 distribution")
plt.tight_layout()

# --- second 2x2: parameter correlations + reduced chi2 ---
fig2, ax2 = plt.subplots(2,2, figsize=(8,8))
ax2[0,0].hist2d(par_a, par_b, bins=30)
ax2[0,0].set_title("b vs a")
ax2[0,1].hist2d(par_a, par_c, bins=30)
ax2[0,1].set_title("c vs a")
ax2[1,0].hist2d(par_b, par_c, bins=30)
ax2[1,0].set_title("c vs b")
ax2[1,1].hist(chi2_red, bins=40)
ax2[1,1].set_title("Reduced chi2")
plt.tight_layout()

# --- chi2 stats ---
ndf = npoints - 3
mean_chi2   = np.mean(chi2_vals)
std_chi2    = np.std(chi2_vals)
exp_mean    = ndf
exp_std     = np.sqrt(2*ndf)
mean_chi2_r = np.mean(chi2_red)

print("Mean chi2 =", mean_chi2, " Expected =", exp_mean)
print("Std chi2  =", std_chi2,  " Expected =", exp_std)
print("Mean reduced chi2 =", mean_chi2_r)

# --- generate one dataset again and plot fit ---
getX(lx)
getY(lx, ly, ley)

A = np.zeros((npoints,3))
for i in range(npoints):
    L = log(lx[i])
    A[i,0] = 1
    A[i,1] = L
    A[i,2] = L*L

W = np.diag(1.0/ley**2)
AtW = A.T @ W
cov = inv(AtW @ A)
p_fit = cov @ (AtW @ ly)

x_smooth = np.linspace(xmin, xmax, 200)
y_fit = p_fit[0] + p_fit[1]*np.log(x_smooth) + p_fit[2]*np.log(x_smooth)**2

# --- save all figures to PDF ---
with PdfPages('LSQFit.pdf') as pdf:
    pdf.savefig(fig1)
    pdf.savefig(fig2)

    # plot the fit with data
    plt.figure(figsize=(8,6))
    plt.errorbar(lx, ly, yerr=ley, fmt='o', label="Data with error bars")
    plt.plot(x_smooth, y_fit, label=f"Fit: a={p_fit[0]:.2f}, b={p_fit[1]:.2f}, c={p_fit[2]:.2f}")
    plt.xlabel("x")
    plt.ylabel("y")
    plt.title("Least Squares Fit on One Dataset")
    plt.legend()
    pdf.savefig()
    plt.close()

    # comments page
    plt.figure()
    plt.axis('off')
    plt.text(0.01, 0.9, "Comments:", fontsize=14)
    plt.text(0.01, 0.8, f"• Mean χ² = {mean_chi2:.2f}  (Expected ≈ {exp_mean})")
    plt.text(0.01, 0.7, f"• Std χ²  = {std_chi2:.2f}  (Expected ≈ {exp_std:.2f})")
    plt.text(0.01, 0.6, f"• Mean reduced χ² = {mean_chi2_r:.2f} (Expected ≈ 1)")
    plt.text(0.01, 0.5, "• More data points → narrower parameter spreads.")
    plt.text(0.01, 0.4, "• Larger sigma → wider spreads and worse χ².")
    pdf.savefig()
    plt.close()

print("Saved LSQFit.pdf")
input("Press Enter to exit")

File: LSQFit.cpp
#include "TRandom2.h"
#include "TGraphErrors.h"
#include "TMath.h"
#include "TApplication.h"
#include "TCanvas.h"
#include "TH2F.h"
#include "TH1F.h"
#include "TGClient.h"
#include "TStyle.h"
#include <iostream>
#include "TF1.h"
#include "TFitResult.h"
#include "TPaveText.h"
#include "TROOT.h"
using namespace std;
using TMath::Log;

//parms
const double xmin=1;
const double xmax=20;
const int npoints=12;
const double sigma=0.2;

double f(double x){
  const double a=0.5;
  const double b=1.3;
  const double c=0.5;
  return a+b*Log(x)+c*Log(x)*Log(x);
}

void getX(double *x){
  double step=(xmax-xmin)/npoints;
  for (int i=0; i<npoints; i++){
    x[i]=xmin+i*step;
  }
}

void getY(const double *x, double *y, double *ey){
  static TRandom2 tr(0);
  for (int i=0; i<npoints; i++){
    y[i]=f(x[i])+tr.Gaus(0,sigma);
    ey[i]=sigma;
  }
}

void leastsq(){
  double x[npoints];
  double y[npoints];
  double ey[npoints];
  getX(x);
  getY(x,y,ey);
  auto tg = new TGraphErrors(npoints,x,y,0,ey);
  tg->Draw("alp");
}

int main(int argc, char **argv){
  TApplication theApp("App", &argc, argv); // init ROOT App for displays

  // ******************************************************************************
  // ** this block is useful for supporting both high and std resolution screens **
  UInt_t dh = gClient->GetDisplayHeight()/2; // fix plot to 1/2 screen height
  //UInt_t dw = gClient->GetDisplayWidth();
  UInt_t dw = 1.1*dh;
  // ******************************************************************************
  gStyle->SetOptStat(0); // turn off histogram stats box

  gROOT->SetBatch(kTRUE);

  TCanvas *tc = new TCanvas("c1","Sample dataset",dw,dh);
  double lx[npoints];
  double ly[npoints];
  double ley[npoints];
  getX(lx);
  getY(lx,ly,ley);
  auto tgl = new TGraphErrors(npoints,lx,ly,0,ley);
  tgl->SetTitle("Pseudoexperiment;x;y"); // An example of one pseudo experiment
  tgl->Draw("alp");
  tc->Draw();

  TH2F *h1 = new TH2F("h1","Parameter b vs a;a;b",100,0,2,100,0,2);
  TH2F *h2 = new TH2F("h2","Parameter c vs a;a;c",100,0,2,100,0,2);
  TH2F *h3 = new TH2F("h3","Parameter c vs b;b;c",100,0,2,100,0,2);
  TH1F *h4 = new TH1F("h4","reduced chi^2;;frequency",100,0,3);

  const int Nexp = 2000;

  TH1F *ha = new TH1F("ha","Parameter a; a; frequency",100,0,2);
  TH1F *hb = new TH1F("hb","Parameter b; b; frequency",100,0,2);
  TH1F *hc = new TH1F("hc","Parameter c; c; frequency",100,0,2);
  TH1F *hchi2 = new TH1F("hchi2","chi^2;;frequency",100,0,60);

  TF1 fitf("fitf","[0] + [1]*log(x) + [2]*log(x)*log(x)", xmin, xmax);

  double sumChi2 = 0.0;
  double sumChi22 = 0.0;

  for (int ie=0; ie<Nexp; ++ie){
    getX(lx);
    getY(lx,ly,ley);
    TGraphErrors tg(npoints,lx,ly,0,ley);
    TFitResultPtr res = tg.Fit(&fitf,"QSN");
    double a = fitf.GetParameter(0);
    double b = fitf.GetParameter(1);
    double c = fitf.GetParameter(2);
    double chi2 = res->Chi2();
    int ndf = res->Ndf();

    ha->Fill(a);
    hb->Fill(b);
    hc->Fill(c);
    hchi2->Fill(chi2);
    h1->Fill(a,b);
    h2->Fill(a,c);
    h3->Fill(b,c);
    if (ndf>0) h4->Fill(chi2/ndf);

    sumChi2  += chi2;
    sumChi22 += chi2*chi2;
  }

  int ndf = npoints - 3;
  double meanChi2 = sumChi2 / Nexp;
  double varChi2  = (sumChi22 / Nexp) - meanChi2*meanChi2;
  if (varChi2 < 0) varChi2 = 0;
  double stdChi2  = TMath::Sqrt(varChi2);
  double expMean  = ndf;
  double expStd   = TMath::Sqrt(2.0*ndf);

  cout << "Mean chi2 = " << meanChi2 << "  Expected = " << expMean << endl;
  cout << "Std chi2  = " << stdChi2  << "  Expected = " << expStd  << endl;
  cout << "Mean reduced chi2 ~ " << meanChi2 / ndf << endl;

  TCanvas *tc1 = new TCanvas("c_dist","parameter/chi2 distributions",200,200,dw,dh);
  tc1->Divide(2,2);
  tc1->cd(1); ha->Draw();
  tc1->cd(2); hb->Draw();
  tc1->cd(3); hc->Draw();
  tc1->cd(4); hchi2->Draw();
  tc1->Draw();

  TCanvas *tc2 = new TCanvas("c2","my study results",200,200,dw,dh);
  tc2->Divide(2,2);
  tc2->cd(1); h1->Draw("colz");
  tc2->cd(2); h2->Draw("colz");
  tc2->cd(3); h3->Draw("colz");
  tc2->cd(4); h4->Draw();
  tc2->Draw();

  tc1->Print("LSQFitROOT.pdf(");
  tc2->Print("LSQFitROOT.pdf");

  TCanvas *tc3 = new TCanvas("c3", "Comments / Analysis", 200, 200, dw, dh);
tc3->cd();
TPaveText *pt = new TPaveText(0.05, 0.05, 0.95, 0.95, "NDC");
pt->SetTextSize(0.03);
pt->AddText("Comments:");
{
    TString s; s.Form("- Mean chi2 = %.2f  (Expected ~ %d)", meanChi2, ndf);
    pt->AddText(s);
}
{
    TString s; s.Form("- Std chi2 = %.2f  (Expected ~ %.2f)", stdChi2, expStd);
    pt->AddText(s);
}
{
    TString s; s.Form("- Mean reduced chi2 = %.2f  (Expected ~ 1)", meanChi2/ndf);
    pt->AddText(s);
}
pt->AddText("- More data points -> narrower parameter spreads.");
pt->AddText("- Larger sigma -> wider parameter spreads and higher chi2.");
pt->Draw();
tc3->Print("LSQFitROOT.pdf)");


  return 0;
}


